### PR TITLE
Add functions to partition uncertainties according to method from Hawkins & Sutton (2009)

### LINKED ISSUE #771: Hawkins-Sutton calculations and graphic
Implement function to create data structure to generate the classic Hawkins-Sutton graphic of relative uncertainty contributions from GHG scenario, model and natural variability. 

- [ ] Lit review
- [ ] Create a test dataset
- [ ] Decide on structure of output data
- [ ] Code the compute function
- [ ] Code the graphic function
- [ ] Tests
- [ ] Documentation
- [ ] Example notebook

# Literature Review

- [Partitioning climate projection uncertainty with multiple large ensembles and CMIP5/6](https://doi.org/10.5194/esd-11-491-2020)
- [A Simple, coherent framework for partitioning uncer- tainty in climate predictions](https://doi.org/10.1175/2011JCLI4085.1)
- [The potential to narrow uncertainty in projections of regional precipitation change](https://doi.org/10.1007/s00382-010-0810-6)
- [The potential to narrow uncertainty in regional climate predictions](https://doi.org/10.1175/2009BAMS2607.1)

### PR DESCRIPTION
**Pull request by huard**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #771 
- [x] Tests for the changes have been added (for bug fixes / features)
  - [ ] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added

### What kind of change does this PR introduce?

* Add functions to compute variance using method from Hawkins & Sutton (2009).

### Does this PR introduce a breaking change?
No

### Other information:
Only tested qualitatively so far. Looks ok but I'd like a more solid validation.
See related PRs:
- https://github.com/Ouranosinc/xclim-testdata/pull/23
- https://github.com/Ouranosinc/spirograph/pull/9

### CODE REVIEW

**Comment by huard** [comment]
Pour voir les résultats, faire rouler le code dans `test_hawkins_sutton`, puis passer la variance à `spirograph.mpl.graph_fraction_of_total_variance`.

File: xclim/ensembles/_partitioning.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+estimates of regional climate uncertainties. Geophysical Research Letters, 45, 926– 934.
+https://doi.org/10.1002/2017GL076297
+
+Lehner, F., Deser, C., Maher, N., Marotzke, J., Fischer, E. M., Brunner, L., Knutti, R., and Hawkins,
+E. (2020). Partitioning climate projection uncertainty with multiple large ensembles and CMIP5/6, Earth Syst. Dynam.,
+11, 491–508, https://doi.org/10.5194/esd-11-491-2020.
+
+Evin, G., Hingray, B., Blanchet, J., Eckert, N., Morin, S., & Verfaillie, D. (2019). Partitioning Uncertainty
+Components of an Incomplete Ensemble of Climate Projections Using Data Augmentation, Journal of Climate, 32(8),
+2423-2440, https://doi.org/10.1175/JCLI-D-18-0606.1
```
Comment (Zeitsperre) [inline xclim/ensembles/_partitioning.py:40]: Do you think you could add these to the BibTeX document and reference these with the `:cite:cts:` directive?

Alternatively, you could specify a mini-bibliography here with the proper configuration. See: https://github.com/Ouranosinc/xclim/blob/8868539166b20b7c95f8c47ca80e9c0226723514/docs/indices.rst

File: xclim/ensembles/_partitioning.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    xr.DataArray, xr.DataArray
+      The mean and the components of variance of the ensemble. These components are coordinates along the
+      `uncertainty` dimension: `variability`, `model`, `scenario`, and `total`.
+
+    References
+    ----------
+    Hawkins, E., & Sutton, R. (2009) The Potential to Narrow Uncertainty in Regional Climate Predictions. Bulletin of
+      the American Meteorological Society, 90(8), 1095–1108. doi:10.1175/2009BAMS2607.1
+    Hawkins, E., and R. Sutton (2011) The potential to narrow uncertainty in projections of regional precipitation
+      change. Climate Dyn., 37, 407–418, doi:10.1007/s00382-010-0810-6.
```
Comment (Zeitsperre) [inline xclim/ensembles/_partitioning.py:91]: Idem.

File: xclim/ensembles/_partitioning.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+Evin, G., Hingray, B., Blanchet, J., Eckert, N., Morin, S., & Verfaillie, D. (2019). Partitioning Uncertainty
+Components of an Incomplete Ensemble of Climate Projections Using Data Augmentation, Journal of Climate, 32(8),
+2423-2440, https://doi.org/10.1175/JCLI-D-18-0606.1
+"""
+
+# Default dimension names
+_dims = dict(model="model", scenario="scenario", member="member")
+
+
+def hawkins_sutton(
```
Comment (Zeitsperre) [inline xclim/ensembles/_partitioning.py:47]: Should `uncertainty_paritioning` be mentioned here? or is Hawkins-Sutton universally known?

Comment (huard) [reply xclim/ensembles/_partitioning.py:47]: There are other algos for uncertainty partitioning, so I want to leave room for them.

File: xclim/ensembles/_partitioning.py (hunk 1 above)
Comment (huard) [reply xclim/ensembles/_partitioning.py:40]: These are within a comment that should not appear in the documentation, and is meant to inform further development. I'll add bibtex items in the bib file, but will leave these comments as is in the file.

File: xclim/ensembles/_partitioning.py (hunk 2 above)
Comment (huard) [reply xclim/ensembles/_partitioning.py:91]: Will do for those.

File: xclim/ensembles/_partitioning.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    # Confirm the same models have data for all scenarios
+    check = da.notnull().any("time").all("scenario")
+    if not check.all():
+        raise ValueError(f"Some models are missing data for some scenarios: \n {check}")
+
+    if weights is None:
+        weights = xr.ones_like(da.model, float)
+
+    # Perform analysis 1950 onward
+    da = da.sel(time=slice("1950", None))
```
Comment (aulemahal) [inline xclim/ensembles/_partitioning.py:119]: Seems strange to me to hardcode this? Or is it a condition from the literature?

**Review by aulemahal — COMMENTED** [review]
Do you think this could be merged without a real test ? What about generating fake data?

File: HISTORY.rst [hunk 5]
```diff
@@ -4,14 +4,17 @@ History
 
 0.41.0 (unreleased)
 -------------------
-Contributors to this version: Trevor James Smith (:user:`Zeitsperre`), Pascal Bourgault (:user:`aulemahal`), Ludwig Lierhammer (:user:`ludwiglierhammer`), Éric Dupuis (:user:`coxipi`).
+
```
[the remark this answers is not in the document]
Comment (huard) [reply HISTORY.rst:7]: Don't know, weird merge behavior? This is an old PR.

File: xclim/ensembles/_filters.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+def model_in_all_scens(da, dimensions=None):
+    """Return data with only simulations that have at least one member in each scenario.
+
+    Parameters
+    ----------
+    da: xr.DataArray
+      Input data with dimensions for time, member, model and scenario.
+    dimensions: dict
+      Mapping from original dimension names to standard dimension names: scenario, model, member.
```
[the remark this answers is not in the document]
Comment (huard) [reply xclim/ensembles/_filters.py:67]: Open to whatever conventions you think is best. I tried to stick with the language in the paper, but I'm not attached to it. The CMIP vocabulary tends to change with each cycle, that's why I went with those. Happy to change to experiment, source and member.

File: xclim/ensembles/_partitioning.py (hunk 4 above)
Comment (huard) [reply xclim/ensembles/_partitioning.py:119]: Yep, it's from the paper. Again, it's the question of how close should these algos stick to the papers they're based on.  No worries adding an argument for this.

File: HISTORY.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply HISTORY.rst:7]: Might have been my bad. Fatfingered?

[the remark this answers is not in the document]
Comment (Zeitsperre) [reply HISTORY.rst:7]:
```suggestion
Contributors to this version: Trevor James Smith (:user:`Zeitsperre`), Pascal Bourgault (:user:`aulemahal`), Ludwig Lierhammer (:user:`ludwiglierhammer`), Éric Dupuis (:user:`coxipi`).
```

**Comment by huard** [comment]
I have data in xclim-testdata for this, but not the expected results. I can only confirm the results make sense by comparing graphics from the papers. Open to ideas on how to go about this.

File: xclim/ensembles/_filters.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (huard) [reply xclim/ensembles/_filters.py:67]: Started to change the naming, and I am getting less convinced it's a good idea. The terms source and experiment make sense to those familiar with the CMIP framework, but for impact modelers, I think models and scenarios is more descriptive.

[the remark this answers is not in the document]
Comment (aulemahal) [reply xclim/ensembles/_filters.py:67]: Ach, yeah ok, let's keep it this way then!

File: xclim/ensembles/_partitioning.py (hunk 4 above)
Comment (aulemahal) [reply xclim/ensembles/_partitioning.py:119]: I think I like this last idea!

File: docs/api.rst [hunk 7]
```diff
@@ -100,6 +100,12 @@ Ensembles Module
 .. autofunction:: xclim.ensembles.robustness_coefficient
    :noindex:
 
+.. automodule:: xclim.ensembles._partitioning
+    :noindex:
+
+.. autofunction:: xclim.ensembles.hawkins_sutton
+    :noindex:
+
```
Comment (aulemahal) [inline docs/api.rst:108]: Should be also document the functions in `ensembles._filters` ?

**Review by aulemahal — APPROVED** [review]
This test seems good!

File: docs/api.rst (hunk 7 above)
Comment (huard) [reply docs/api.rst:108]: I wasn't sure they were ready for prime-time. The tests for those are fairly basic.

File: xclim/ensembles/_filters.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        +============+==================================+
+        | historical | hhhhhhhhhhhhhhhh                 |
+        +------------+----------------------------------+
+        | ssp245     |                 111111111111     |
+        +------------+----------------------------------+
+        | ssp370     |                 222222222222     |
+        +------------+----------------------------------+
+
+    to:
+        +----------+----------------------------------+
```
Comment (Zeitsperre) [inline xclim/ensembles/_filters.py:37]: Something strange going on with the formatting here: 
image

File: CHANGES.rst [hunk 9]
```diff
@@ -1365,7 +1369,7 @@ v0.15.x (2020-03-12)
 --------------------
 * Improvement in FWI: Vectorization of DC, DMC and FFMC with numba and small code refactoring for better maintainability.
 * Added example notebook for creating a catalog of selected indices
-* Added `growing_season_end`, `last_spring_frost`, `dry_days`,  `hot_spell_frequency`, `hot_spell_max_length`, and `maximum_consecutive_frost_free_days` indices.
+* Added `growing_season_end`, `last_spring_frost`, `dry_days`, `hot_spell_frequency`, `hot_spell_max_length`, and `maximum_consecutive_frost_free_days` indices.
```
Comment (Zeitsperre) [inline CHANGES.rst:1372]: Was this automatically adjusted by a pre-commit hook? Or did you happen upon this and fix it manually?

File: CHANGES.rst [hunk 10]
```diff
@@ -2,6 +2,10 @@
 Changelog
 =========
 
+v0.45.0 (unreleased)
+--------------------
+* ``ensembles.hawkins_sutton`` method to partition the uncertainty sources in a climate projection ensemble. (:issue:`771`, :pull:`1262`).
```
Comment (Zeitsperre) [inline CHANGES.rst:7]:
```suggestion
Contributors to this version: David Huard (:user:`huard`).

New features and enhancements
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
* ``ensembles.hawkins_sutton`` method to partition the uncertainty sources in a climate projection ensemble. (:issue:`771`, :pull:`1262`).
```

File: CHANGES.rst (hunk 9 above)
Comment (huard) [reply CHANGES.rst:1372]: Don't know, but I didn't modify CHANGES recently.

File: xclim/ensembles/_partitioning.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+
+from __future__ import annotations
+
+import numpy as np
+import pandas as pd
+import xarray as xr
+
+"""
+Implemented partitioning algorithms:
```
[the remark this answers is not in the document]
Comment (huard) [reply xclim/ensembles/_partitioning.py:18]: Maybe when there's more than one implementation, then yes.

File: CHANGES.rst (hunk 9 above)
Comment (Zeitsperre) [reply CHANGES.rst:1376]: Haha, wow. I changed this in March. Apologies. :facepalm:

**Pull request merged**